### ISSUE TITLE
Default methods Content-Type headers cant be overridden

### ISSUE THREAD

**Issue by Ashot-KR** [axios/axios#2623]
**Describe the bug**
When setting default `Content-Type` for methods (e.g. `patch`) always sends `ContentType: application/json;charset=utf-8`

**To Reproduce**
```js
const api = axios.create({
  baseURL: '/api',
  headers: {
    common: {
      Accept: 'application/json'
    },
    patch: {
      'Content-Type': 'application/merge-patch+json'
    }
  }
})

api.patch('/user/1', { foo: 'bar' })
```

**Expected behavior**
Should send `Content-Type: application/merge-patch+json` header.

**Environment:**
 - Axios Version 0.19.0
 - OS: macOS 10.15.1
 - Browser Chrome, Safari
 - Browser Version 78.0.3904.108

**Additional context/Screenshots**
I think [problem is here](https://github.com/axios/axios/blob/master/lib/core/dispatchRequest.js#L40) – root headers from `config.headers` always overrides headers for specific methods. In example above `config.headers['Content-Type']` not set, so axios set it to `application/json;charset=utf-8` under the hood and `config.headers['Content-Type']` overrides `config.headers.patch['Content-Type']`.
I think methods headers should have higher priority and should be passed to `utils.merge` function in last argument

**Comment by davidlandais** [comment]
@Ashot-KR I've got the same problem. As a workaround, i'm using an interceptor.
```
axios.interceptors.request.use((config) => {
    const headers = merge({}, config.headers.common, config.headers.patch);
    return merge({}, config, { headers });
  })
```
It is working for me. Hope this help.

**Comment by pelepelin** [comment]
I've found a similar issue, i think it's the same, but my analysis is different.

**Issue**

When an object is passed in `data` and Axios default `transformRequest` is not overridden, the `Content-Type` configured for a method is not used, instead, the `Content-Type` is always `application/json;charset=utf-8`.

**Reproduce**
```js
require('axios').create({
  headers: {
    post: {
      'Content-Type': 'application/json'
    }
  }
}).request({
  url: 'http://httpbin.org/anything',
  method: 'post',
  data: {}
}).then(response => console.log(response.data.headers['Content-Type']));
```
Result:
```
application/json;charset=utf-8
```

**Reason**

`transformRequest` is applied before merging headers:
https://github.com/axios/axios/blob/351cf290f0478d6e47e74c6da2f3ad8fe8f29887/lib/core/dispatchRequest.js#L30-L37. 

However, default `transformRequest` sets the `Content-Type` if it's not set:
https://github.com/axios/axios/blob/351cf290f0478d6e47e74c6da2f3ad8fe8f29887/lib/defaults.js#L50-L51

So, default by method `Content-Type` is not applied.

**Comment by pelepelin** [comment]
**Workaround**

Apply a custom `transformRequest` suitable for your usecase instead.

```js
require('axios').create({
  transformRequest: (data, headers) => {
    if (data !== undefined) {
      headers['Content-Type'] = 'application/json';

      return JSON.stringify(data)
    } else {
      return data;
    }
  }
}).request({
  url: 'http://httpbin.org/anything',
  method: 'post',
  data: {}
}).then(response => console.log(response.data.headers['Content-Type']));
```

Result:
```
application/json
```

**Comment by pelepelin** [comment]
**Workaround 2**

Serialize data explicitly.

```js
require('axios').create({
  headers: {
    post: {
      'Content-Type': 'application/json'
    }
  },
}).request({
  url: 'http://httpbin.org/anything',
  method: 'post',
  data: JSON.stringify({})
}).then(response => console.log(response.data.headers['Content-Type']));
```

Result:

```
application/json
```

**Comment by luisfuertes** [comment]
+1

**Comment by jasonsaayman** [comment]
Fixed in #2293

**Comment by santerref** [comment]
A solution that is working for me is to set the Content-Type header in the headers property instead of common:

```
axios.defaults.headers['Content-Type'] = 'application/vnd.api+json';
```

I think the issue is in dispatchRequest.js:

```javascript
// Transform request data
config.data = transformData(
  config.data,
  config.headers,
  config.transformRequest
);

// Flatten headers
config.headers = utils.merge(
  config.headers.common || {},
  config.headers[config.method] || {},
  config.headers
);
```

The flatten headers should be done BEFORE the transform request data. Otherwise, axios set the Content-Type to application/json;utf-8 if the data sent is an object and after the merge cannot override it since it is set.

Also, the function setContentTypeIfUnset should have a third parameter to look into the current method of the request or have common as default value if not provided:

```javascript
function setContentTypeIfUnset(headers, value) {
  if (!utils.isUndefined(headers) && utils.isUndefined(headers['Content-Type'])) {
    headers['Content-Type'] = value;
  }
}
```

Currently, it looks for headers['Content-Type'] and not headers.common['Content-Type'].

**Comment by mariukhin** [comment]
Just change the format of the sent data to JSON
axios can't recognize the data type of sended data, so it add 'charset=utf-8'

Use stringify
`JSON.stringify(your data)`

**Comment by WalderlanSena** [comment]
**is the last parameter for me is working, for example:**

```javascript
const response = api.patch, url + '/' + id, yourData, {
  headers: {
      'Accept': 'application/hal+json',
      'Content-Type': 'application/merge-patch+json'
  }
};
```

**Comment by pelepelin** [comment]
@WalderlanSena your example uses explicit headers settings for the request. This issue is about defaults provided to `axios.create`.

**Comment by jasonsaayman** [comment]
I believe this should be amended in the latest version of Axios, please try it out and if the issue still exists, please open a new pull request.

**Comment by jraller** [comment]
@jasonsaayman which version was this fixed in? I'm seeing it in 0.21.1 (the version at the time of your last comment) which is used by the nuxt-community axios-module and would like to know if upgrading that and nuxt auth to use 0.24 would allow me to set a default Content-Type header for my patch calls. I don't see this issue listed in the change log, or an entry that appears to match.

I looked in the tests and I'm not seeing test coverage for this case. Did I miss it?

**Comment by renevall** [comment]
In my case I had to do this:

```typescript
this.instance.defaults.transformRequest = [(data, headers) => {
            if (data && headers) {
                if (headers["content-type"] === "application/json" || headers["Content-Type"] === "application/json") {
                    headers["Content-Type"] = "application/json"
                    return JSON.stringify(snakecaseKeys(data, { deep: true }))
                }
            }
        }]
```

I think there is an inconsistency going on, because response sees is as "content-type" and request as "Content-Type". If you try in your instance creation
```typescript
this.instance = axios.create({
            baseURL,
            headers: { 'content-type': 'application/json' }

        });

```

Because of the merge process, then axios will send content-type: "application/json, application/json". Feel this is error prone.
